### PR TITLE
Upgrading drawer

### LINKED ISSUE #1251: 2.0 - Add gesture support to drawer
We need to add support for swipe gesture to our modal drawer, because its 2019.

### PR DESCRIPTION
**Pull request by prateekbh**
Fixes #1251 #1250 

- Adds gesture support for drawer
- Upgrades Drawer
- Confirms to #1293 

PMCSlider

### CODE REVIEW

**Comment by cromefire** [comment]
Nice one

File: packages/drawer/src/index.tsx [hunk 1]
```diff
@@ -15,11 +16,32 @@ export interface IDrawerProps extends JSX.HTMLAttributes {
...
+export interface IDrawerState {
+  sliderState: SLIDER_STATES;
+  touchPositionX: number;
+  touchPositionY: number;
+  swipeDistance: number;
+}
+
+const CLOSED_STATE = {
+  sliderState: SLIDER_STATES.REST,
+  swipeDistance: -256,
```
Comment (cromefire) [inline packages/drawer/src/index.tsx]: Is this okay to be hardcoded?

File: packages/drawer/src/index.tsx [hunk 2]
```diff
@@ -55,13 +82,105 @@ export class Drawer extends MaterialComponent<IDrawerProps, IDrawerState> {
...
+          {...props}
+          style={styleObject}>
           {props.children}
         </aside>
         {props.modal && <div class="mdc-drawer-scrim" />}
       </div>
     );
   }
+
+  private handleTouchStart_ = e => {
```
Comment (cromefire) [inline packages/drawer/src/index.tsx]: Why do they all have an underscore at the end?

File: packages/drawer/src/index.tsx [hunk 3]
```diff
@@ -55,13 +82,105 @@ export class Drawer extends MaterialComponent<IDrawerProps, IDrawerState> {
...
+    const startPosition = this.MDComponent.open ? 0 : -256;
+    const {clientX, clientY} = touches[0];
+    const deltaX = clientX - this.state.touchPositionX;
+    this.setState({
+      sliderState: SLIDER_STATES.SWIPING,
+      swipeDistance: startPosition + deltaX
+    });
+  };
+
+  private attachGlobalListeners_ = () => {
```
Comment (cromefire) [inline packages/drawer/src/index.tsx]: Are they cleaned up again?

**Comment by cromefire** [comment]
Is there any way that we can automatically test this? I guess we'll have to look into drag & drop support

**Comment by prateekbh** [comment]
@cromefire yeah we just need to enable this for mousedown/up/move and we'll be able to test it with cypress

**Comment by prateekbh** [comment]
@cromefire Tests for swipe added

**Comment by prateekbh** [comment]
ping

**Comment by cromefire** [comment]
There are still 3 comments left unanswered

File: packages/drawer/src/index.tsx (hunk 1 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: we dont really have a good way of detecting this... so our best bet are our visual tests

File: packages/drawer/src/index.tsx (hunk 2 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: cuz they are private and we ship js not TS. WDYT?

File: packages/drawer/src/index.tsx (hunk 3 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: missed that

File: packages/drawer/src/index.tsx (hunk 2 above)
Comment (cromefire) [reply packages/drawer/src/index.tsx]: Most IDEs use the .d.ts file also to suggest things in JavaScript (at least WebStorm and VSCode I think)

File: packages/drawer/src/index.tsx (hunk 1 above)
Comment (cromefire) [reply packages/drawer/src/index.tsx]: We will see

File: packages/drawer/src/index.tsx (hunk 2 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: i know but 
a. lets not take that for granted
b. does it hurt? even MDC-web ships this way

Comment (cromefire) [reply packages/drawer/src/index.tsx]: It's not consistent with the rest of the codebase and I think it's quite ugly, but if you really want to do this

**Comment by cromefire** [comment]
Also since when do we have netlify?

**Comment by cromefire** [comment]
I think after this there is not that much to do until we can roll the first alpha 🎉

**Comment by cromefire** [comment]
This does not look correct:

Screenshot

**Comment by prateekbh** [comment]
yep thats y added a preview deploy from netlify also the navigation links are not really working

File: packages/drawer/src/index.tsx (hunk 1 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: yeah i really dont have a good way of calculating this right now

File: packages/drawer/src/index.tsx (hunk 2 above)
Comment (prateekbh) [reply packages/drawer/src/index.tsx]: i'll remove it

**Comment by prateekbh** [comment]
@cromefire all known corner cases are now covered.

I'll work on some cross browser issues post alpha release

**Comment by cromefire** [comment]
That looks quite goo but let us look at 2
Short things:
- The drawer is only visible if I Touch it
- I think we should extract that into it's own class instead using a prop (can do that)

**Comment by prateekbh** [comment]
Gotcha... Will do in next release

**Pull request merged**